pkg/controller/kubedirectorcluster/kubedirectorcluster_controller.go

```diff
@@ -0,0 +1,126 @@
+var KDCReconciler *ReconcileKubeDirectorCluster
+
+// newReconciler returns a new reconcile.Reconciler .
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	if KDCReconciler == nil {
+		KDCReconciler = &ReconcileKubeDirectorCluster{
+			Client: mgr.GetClient(),
+			scheme: mgr.GetScheme(),
+			lock:   sync.RWMutex{},
+			clusterState: reconcilerClusterState{
+				clusterStatusGens: make(map[types.UID]StatusGen),
+				clusterAppTypes:   make(map[string]string),
+			},
+		}
+	}
+	return KDCReconciler
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("kubedirectorcluster-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource KubeDirectorCluster
+	err = c.Watch(&source.Kind{Type: &v1alpha1.KubeDirectorCluster{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileKubeDirectorCluster implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileKubeDirectorCluster{}
+
+// ReconcileKubeDirectorCluster reconciles a KubeDirectorCluster object
+type ReconcileKubeDirectorCluster struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	Client       client.Client
+	scheme       *runtime.Scheme
+	lock         sync.RWMutex
+	clusterState reconcilerClusterState
+}
+
+// Reconcile reads that state of the cluster for a KubeDirectorCluster object and makes changes based
+// on the state read and what is in the KubeDirectorCluster.Spec
+//
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileKubeDirectorCluster) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling KubeDirectorCluster")
+
+	// Fetch the KubeDirectorCluster instance
+	kdCluster := &v1alpha1.KubeDirectorCluster{}
+	err := r.Client.Get(context.TODO(), request.NamespacedName, kdCluster)
+	if err != nil {
+		// If the resource is not found, that means all of
+		// the finalizers have been removed, and the kubedirectorcluster
+		// resource has been deleted, so there is nothing left
+		// to do.
+		if apierrors.IsNotFound(err) {
+			// TODO: I put this code here temporarily to make the behaviour
+			//  of the existing code, but we don't have kdCluster's UID at
+			//  this point because the cluster has been deleted.
+			//  deleteStatusGen() needs to be called or we will leak entries
+			//  in the map.
+			// deleteStatusGen(kdCluster, r)
```

Comment:
Addressed in my long comment.

I also wondered whether we should move the invocation of removeClusterAppReference. Not sure? It's kind of nice that in this case we can do this cleanup when the object is for-sure-for-sure deleted, but it might be best to keep the cleanup activities bundled together.

If we did move it to handleFinalizers, then we would also have to modify syncCluster a bit so that it doesn't do ensureClusterAppReference until AFTER any early exit triggered by handleFinalizers.

Reply:
I decided to move the  invocation of removeClusterAppReference to handleFinalizers() to, as you said, keep all of the cleanup activities bundled together. I think that's cleaner and easier to understand.

I don't think we have to move the call to  ensureClusterAppReference after the early exit triggered by handleFinalizers. Currently we'll  call ensureClusterAppReference at the beginning of syncCluster, then we'll call removeClusterAppReference from handleFinalizers and then exit early from syncCluster. Is there an issue with that sequence?

Reply:
I just read my previous reply and I feel I need to expand it a little.

I don't think we have to move the call to ensureClusterAppReference after the early exit triggered by handleFinalizers. Currently we'll call ensureClusterAppReference at the beginning of syncCluster that potentially adds an app reference for the KD cluster, then we'll call removeClusterAppReference from handleFinalizers that removes the KD Cluster's apps reference, and then we exit early from syncCluster. Bottom line is that the reference is always removed when the KD cluster is deleted which is what we want.

Is there an issue with that sequence?

Reply:
Seems fine yeah. Trying to remember if there was something else that bothered me about that beyond the potential redundancy... I guess not?